File: neetcode/14. 2-d dynamic programming/16.numberOfWaysToRearrangeSticksWithKSticksVisible.cpp
// There are n uniquely-sized sticks whose lengths are integers from 1 to n. 
// You want to arrange the sticks such that exactly k sticks are visible from the left. 
// A stick is visible from the left if there are no longer sticks to the left of it.

// For example, if the sticks are arranged [1,3,2,5,4], then the sticks with lengths 1, 3, and 5 
// are visible from the left.
// Given n and k, return the number of such arrangements. Since the answer may be large, 
// return it modulo 109 + 7.

#include <vector>
using namespace std;

// top-down dp approach.
// Time complexity: O(n * k)
#define ll long long
class Solution {
public:
  vector<vector<ll> > dp;
  ll mod = 1e9 + 7;
  
  ll numWays(ll n, ll k) {
    if(k == 0 || k > n) return 0; // can't have no stick visible or more sticks visible than the total number of sticks
    if(n <= 2) return 1; // base condition
    
    if(dp[n][k] != -1) return dp[n][k]; // return if already calculated
    
    ll ans = 0;
    
    // select the tallest stick
    ans = (ans + numWays(n - 1, k - 1)) % mod;
    
    // select any of the n - 1 sticks that are not the tallest
    ans = (ans + ((n - 1) * numWays(n - 1, k))) % mod;
    
    return dp[n][k] = ans;
  }
  
  int rearrangeSticks(int n, int k) {
    dp = vector<vector<ll> >(n + 1, vector<ll>(k + 1, -1));
    return numWays(n, k);
  }
};

// C++ (4 ms)
// The runtime is 4 ms because our dp is global, thus re-used by all test cases.
// Inversion approach
int dp[1001][1001] = {}, mod = 1000000007;
class Solution {
public:
  int rearrangeSticks(int n, int k) {
    if (k == n)
      return 1;
    if (k > 0 && n > k && !dp[n][k])
      dp[n][k] = ((long)(n - 1) * rearrangeSticks(n - 1, k) + rearrangeSticks(n - 1, k - 1) + dp[n][k]) % mod;
    return dp[n][k];
  }
};

File: neetcode/14. 2-d dynamic programming/16.numberOfWaysToRearrangeSticksWithKSticksVisible.py
# There are n uniquely-sized sticks whose lengths are integers from 1 to n. 
# You want to arrange the sticks such that exactly k sticks are visible from the left. 
# A stick is visible from the left if there are no longer sticks to the left of it.

# For example, if the sticks are arranged [1,3,2,5,4], then the sticks with lengths 1, 3, and 5 
# are visible from the left.
# Given n and k, return the number of such arrangements. Since the answer may be large, 
# return it modulo 109 + 7.

from math import comb, perm
from functools import cache

M = 10 ** 9 + 7
class Solution:
    def rearrangeSticks(self, n: int, k: int) -> int:
        return self.dp(n, k)

    # consider the shortest
    @cache
    def dp(self, n, k):
        if n == k:
            return 1
        if n <= 0 or k <= 0:
            return 0
        return (self.dp(n - 1, k - 1) + (n - 1) * self.dp(n - 1, k)) % M  
# `dp(n, k)` indicates the number of arrangements of `n` distinct sticks with exactly `k` visible sticks. 
# Those `n` **distinct** sticks does NOT have to be numbered from `1` to `n` consecutively.


## Python Math Formula, no DP, 100% runtime
class Solution:
    def rearrangeSticks(self, n: int, k: int) -> int:
        MOD = 10 ** 9 + 7
        # Precalculate all necessary modular inverses and factorial inverses from 1 to 2 * n - k
        upper_bound_precalc = 2 * n - k + 2
        inverses = [1] * (upper_bound_precalc + 1)
        inverse_factorials = [1] * (upper_bound_precalc + 1)

        for a in range(2, upper_bound_precalc + 1):
            inverses[a] = MOD - MOD // a * inverses[MOD % a] % MOD
            inverse_factorials[a] = (inverses[a] * inverse_factorials[a - 1]) % MOD

        # Precompute exponents
        total = 0
        my_pow_list = [pow(j, n - k, MOD) for j in range(n - k + 1)]

        for i in range(n - k + 1):
            multiplier = (inverses[n + i] * inverse_factorials[n - k - i] * inverse_factorials[n - k + i]) % MOD
            add_here = 0

            # Follow the math formula
            for j in range(i + 1):
                if j % 2 == 0:
                    add_here += my_pow_list[j] * inverse_factorials[j] * inverse_factorials[i - j]
                else:
                    add_here -= my_pow_list[j] * inverse_factorials[j] * inverse_factorials[i - j]

            # Update the powers list
            my_pow_list = [(j * my_pow_list[j]) % MOD for j in range(n - k + 1)]

            total += multiplier * add_here % MOD

        # Multiply by factorials, and flip the sign of answer if n != k mod 2
        total %= MOD
        total *= pow(inverse_factorials[2 * n - k], -1, MOD) * inverse_factorials[k - 1] * ((-1) ** ((n - k) % 2))

        return total % MOD
    

## Using Fast fourier transform for polynomial multiplication in O(n log n) time
import numpy as np
from scipy.signal import convolve

class Solution:
    def rearrangeSticks(self, n: int, k: int) -> int:
        MOD = 10 ** 9 + 7
        
        # FFT multiply
        def multiply_2_polynomials(poly1, poly2):
            return convolve(poly1, poly2) % MOD

        def mult_poly_list(poly_list):
            if len(poly_list) == 1:
                return poly_list[0]

            if len(poly_list) == 2:
                return multiply_2_polynomials(poly_list[0], poly_list[1])

            return multiply_2_polynomials(mult_poly_list(poly_list[:len(poly_list) // 2]),
                                          mult_poly_list(poly_list[len(poly_list) // 2:]))

        # Generate all solutions for n: j=0,1,...n. Return coefficient at k
        def get_coeff(want_coeff):
            my_polys = []
            for x in range(n):
                my_polys.append(np.array([x, 1], dtype=object))

            return mult_poly_list(my_polys)[want_coeff]

        if n == k:
            return 1

        return get_coeff(k)
    